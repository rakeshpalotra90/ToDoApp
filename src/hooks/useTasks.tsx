import React, {
  useContext,
  createContext,
  useState,
  useMemo,
  useEffect,
} from "react";
import gql from "graphql-tag";
import { useApolloClient, useQuery } from "@apollo/client";

const taskContext = createContext<TasksProvider | null>(null);

export const useTasks = () => useContext<TasksProvider | null>(taskContext);

export function ProvideTasks({ children }: any) {
  const value = _useTasks();
  return <taskContext.Provider value={value}>{children}</taskContext.Provider>;
}

type Tags = {
  id: number;
  title: string;
}
export interface Task {
  id: number;
  text: string;
  completed: boolean;
  priority: boolean;
  repeat: boolean;
  tags: Tags[];
  timeStamp: number | null;
  comment: number;
};

interface TasksProvider {
  tasks: Task[];
  loading: boolean;
  toggleTodo: (id: number) => void;
}

function _useTasks() {
  const [tasks, setTasks] = useState<Task[]>([]);
  const apollo = useApolloClient();
  const { data, loading, error } = useQuery(GET_ALL_TODOS);

  useEffect(() => {
    if (!data?.todos) return;
    setTasks(data?.todos ?? []);
  }, [data]);

  const toggleTodo = (id: number) => {
    return apollo.mutate({
      mutation: TOGGLE_TODO,
      variables: {
        id,
      },
    });
  };

  return useMemo(
    () => ({
      tasks,
      loading,
      toggleTodo,
    }),
    [tasks, loading]
  );
}

export const GET_ALL_TODOS = gql`
  query GetAllTodos {
    todos @client {
      id
      text
      completed
      priority
      repeat
      tags
      timeStamp
      comment
    }
  }
`;

const TOGGLE_TODO = gql`
  mutation ToggleTodo($id: Int!) {
    toggleTodo(id: $id) @client
  }
`;