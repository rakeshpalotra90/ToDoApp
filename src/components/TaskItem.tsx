import React, { useMemo } from "react";
import { StyleSheet, View, Text, TouchableOpacity } from "react-native";
import moment from "moment";

import DefaultTheme from "../constants/theme/default";
import Svgs from "../constants/svgs";

import { Task } from "../hooks/useTasks";

function getTagBackgroung(title: string) {
  switch (title) {
    case "Celebration":
        return DefaultTheme.tag15;
    case "Chores":
      return DefaultTheme.tag18;
    default:
      return DefaultTheme.tag17;
  }
}

const TaskItem = ({
  data,
  toggleTodo,
}: {
  data: Task;
  toggleTodo: (id: number) => void;
}) => {
  return useMemo(
    () => (
      <View style={styles.itemContainer}>
        <View style={styles.itemRow}>
          <View style={styles.itemColumnContainer}>
            <View style={styles.itemRowContainer}>
              {data.priority && (
                <View style={styles.itemIconLeftContainer}>
                  <Svgs.Priority />
                </View>
              )}
              <View style={styles.itemTitleContent}>
                <Text
                  style={[
                    styles.itemTitle,
                    data.completed && styles.lineThrough,
                  ]}
                >
                  {data.text}
                </Text>
              </View>
            </View>
            {data.timeStamp && (
              <View style={styles.itemRowContainer}>
                <View style={styles.itemIconTextContainer}>
                  <Svgs.Date />
                  <Text style={styles.itemIconText}>
                    {moment(data.timeStamp).format("DD MMM")}
                  </Text>
                </View>
                <View style={styles.itemIconTextContainer}>
                  <Svgs.Time />
                  <Text style={styles.itemIconText}>
                    {moment(data.timeStamp).format("HH.mm")}
                  </Text>
                </View>
                <View style={styles.itemIconTextContainer}>
                  <Svgs.Comment />
                  <Text style={styles.itemIconText}>{data.comment}</Text>
                </View>
                {data.repeat && (
                  <View style={styles.itemIconTextContainer}>
                    <Svgs.Repeat />
                  </View>
                )}
              </View>
            )}
            {(data.tags ?? []).length > 0 && (
              <View key="tags" style={styles.itemRowContainer}>
                {data.tags.map((item, index) => {
                  return (
                    <View
                      key={index}
                      style={[
                        styles.itemTagContainer,
                        { backgroundColor: getTagBackgroung(item.title) },
                      ]}
                    >
                      <Text style={styles.itemTagText}>{item.title}</Text>
                    </View>
                  );
                })}
              </View>
            )}
          </View>
          <View style={styles.itemIconRightContainer}>
            <TouchableOpacity key={'completedId'} onPress={() => toggleTodo(data.id)}>
              {data.completed ? <Svgs.RadioActive /> : <Svgs.RadioInactive />}
            </TouchableOpacity>
          </View>
        </View>
      </View>
    ),
    [data]
  );
};

const styles = StyleSheet.create({
  itemContainer: {
    margin: 5,
    paddingHorizontal: 14,
    paddingVertical: 10,
    borderColor: DefaultTheme.grey20,
    borderWidth: 1,
    borderRadius: 9,
  },
  itemRow: {
    flexDirection: "row",
  },
  itemRowContainer: {
    flexDirection: "row",
    alignItems: "center",
    flexWrap: "wrap",
  },
  itemTitleContent: {
    flex: 1,
  },
  itemTitle: {
    fontSize: 16,
    lineHeight: 20,
    color: DefaultTheme.black,
    fontWeight: "700",
  },
  lineThrough: {
    textDecorationLine: "line-through",
  },
  itemIconLeftContainer: {
    paddingRight: 5,
  },
  itemIconRightContainer: {
    paddingLeft: 5,
  },
  itemColumnContainer: {
    flex: 1,
  },
  itemIconTextContainer: {
    marginTop: 8,
    marginRight: 10,
    flexDirection: "row",
    alignItems: "center",
  },
  itemIconText: {
    fontSize: 13,
    color: DefaultTheme.grey90,
    fontWeight: "400",
    paddingLeft: 4,
  },
  itemTagContainer: {
    marginTop: 8,
    marginRight: 4,
    paddingHorizontal: 8,
    paddingVertical: 4,
    borderRadius: 4,
    backgroundColor: DefaultTheme.tag15,
  },
  itemTagText: {
    fontSize: 12,
    color: DefaultTheme.black,
    fontWeight: "400",
  },
});

export default TaskItem;
