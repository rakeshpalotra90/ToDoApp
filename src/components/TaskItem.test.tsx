import React from "react";
import { shallow } from "enzyme";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Task } from "../hooks/useTasks";

import TaskItem from "./TaskItem";

const data: Task = {
  id: 1,
  text: "Call carpenter to fix kitchen pipe",
  completed: false,
  priority: true,
  repeat: false,
  tags: [
    {
      id: 1,
      title: "Home",
    },
    {
      id: 2,
      title: "Celebration",
    },
    {
      id: 3,
      title: "Chores",
    },
  ],
  timeStamp: null,
  comment: 0,
};

Enzyme.configure({ adapter: new Adapter() });

describe("<TaskItem />", () => {
  const toggleTodo = jest.fn();
  const wrapper = shallow(<TaskItem data={data} toggleTodo={toggleTodo} />);

  it("renders correctly", () => {
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it("should have tags", () => {
    const tag = wrapper.findWhere((n) => n.key() === `tags`);
    tag.contains("Home");
  });

  it("should call toggleTodo", () => {
    const button = wrapper.findWhere((n) => n.key() === `completedId`);
    button.simulate("press");
  });
});
