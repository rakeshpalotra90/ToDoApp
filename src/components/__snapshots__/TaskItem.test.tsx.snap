// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskItem /> renders correctly 1`] = `ShallowWrapper {}`;
